catalog/core/catalog-core-validator/src/main/java/ddf/catalog/validation/impl/validator/RelationshipValidator.java

```diff
@@ -0,0 +1,257 @@
+import java.util.stream.Collectors;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang.StringUtils;
+
+public class RelationshipValidator implements ReportingMetacardValidator, MetacardValidator {
+  private static final String MUST_HAVE = "mustHave";
+  private static final String CANNOT_HAVE = "cannotHave";
+  private static final String CAN_ONLY_HAVE = "canOnlyHave";
+  private final Map<String, Function<Attribute, Optional<ValidationViolation>>> relationships =
+      new HashMap<>();
+  private final String sourceAttribute;
+  private final String sourceValue;
+  private final String relationship;
+  private final String targetAttribute;
+  private final List<String> targetValues;
+
+  public RelationshipValidator(
+      String sourceAttribute,
+      String sourceValue,
+      String relationship,
+      String targetAttribute,
+      String... targetValues) {
+    relationships.put(MUST_HAVE, this::mustHave);
+    relationships.put(CANNOT_HAVE, this::cannotHave);
+    relationships.put(CAN_ONLY_HAVE, this::canOnlyHave);
+    this.sourceAttribute = sourceAttribute;
+    this.sourceValue = sourceValue;
+    this.relationship = relationship;
+    this.targetAttribute = targetAttribute;
+    if (targetValues != null) {
+      this.targetValues =
+          Arrays.stream(targetValues)
+              .filter(Objects::nonNull)
+              .map(Objects::toString)
+              .collect(Collectors.toList());
+    } else {
+      this.targetValues = Collections.emptyList();
+    }
+    if (!relationships.keySet().contains(relationship)) {
+      throw new IllegalArgumentException(
+          "Unrecognized relationship " + relationship + " for validator " + this.toString());
+    }
+    if (CAN_ONLY_HAVE.equals(relationship) && CollectionUtils.isEmpty(this.targetValues)) {
+      throw new IllegalArgumentException(
+          "Relationship " + CAN_ONLY_HAVE + " must specify one or more target values");
+    }
+  }
+
+  @Override
+  public Optional<MetacardValidationReport> validateMetacard(Metacard metacard) {
+    Attribute attribute = metacard.getAttribute(sourceAttribute);
+    if (attribute != null
+        && attribute.getValues().stream().anyMatch(Objects::nonNull)
+        && (StringUtils.isEmpty(sourceValue) || attribute.getValues().contains(sourceValue))) {
+      MetacardValidationReportImpl report = new MetacardValidationReportImpl();
+      Optional<ValidationViolation> violation =
+          relationships.get(relationship).apply(metacard.getAttribute(targetAttribute));
+      violation.ifPresent(report::addAttributeViolation);
+      if (violation.isPresent()) {
+        return Optional.of(report);
+      } else {
+        return Optional.empty();
+      }
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  private Optional<ValidationViolation> mustHave(Attribute attribute) {
+    return getValidationViolation(
+        "must have",
+        attribute == null
```

Comment:
✏️ Could we at least create some local variables for these boolean expressions to reduce nesting?